Document/CategoryManager.php

```diff
@@ -43,4 +158,85 @@ public function getPager(array $criteria, $page, $limit = 10, array $sort = arra
 
         return $pager;
     }
+
+    /**
+     * Load all categories from the database, the current method is very efficient for < 256 categories.
+     */
+    protected function loadCategories(ContextInterface $context)
+    {
+        if (array_key_exists($context->getId(), $this->categories)) {
+            return;
+        }
+
+        $categories = $this->getObjectManager()->createQuery(sprintf('SELECT c FROM %s c WHERE c.context = :context ORDER BY c.parent ASC', $this->getClass()))
+            ->setParameter('context', $context->getId())
+            ->execute();
+
+        if (count($categories) == 0) {
+            // no category, create one for the provided context
+            $category = $this->create();
+            $category->setName($context->getName());
+            $category->setEnabled(true);
+            $category->setContext($context);
+            $category->setDescription($context->getName());
+
+            $this->save($category);
+
+            $categories = array($category);
+        }
+
+        foreach ($categories as $pos => $category) {
+            if ($pos === 0 && $category->getParent()) {
+                throw new \RuntimeException('The first category must be the root');
+            }
+
+            if ($pos === 0) {
+                $root = $category;
+            }
+
+            $this->categories[$context->getId()][$category->getId()] = $category;
+
+            $parent = $category->getParent();
+
+            $category->disableChildrenLazyLoading();
+
+            if ($parent) {
+                $parent->addChild($category);
+            }
+        }
+
+        $this->categories[$context->getId()] = array(
+            0 => $root,
+        );
+    }
+
+    /**
+     * @param $contextCode
+     *
+     * @return ContextInterface
+     */
+    private function getContext($contextCode)
+    {
+        if (empty($contextCode)) {
+            $contextCode = ContextInterface::DEFAULT_CONTEXT;
+        }
+
+        if ($contextCode instanceof ContextInterface) {
```

Comment:
If `$contextCode` instanceof ContextInterface, should it really be named `$contextCode` ?